File: le-serviceflows/core/src/main/resources/com/latticeengines/serviceflows/core/transforms/std_visidb_ds_spamindicator.py

#
# $LastChangedBy$
# $LastChangedDate$
# $Rev$
#

import re
from std_visidb_ds_firstname_sameas_lastname import std_visidb_ds_firstname_sameas_lastname
from std_visidb_ds_companyname_entropy import std_visidb_ds_companyname_entropy
from std_entropy import std_entropy

def metadata():
    return {  'ApprovedUsage'   : 'ModelAndModelInsights'
            , 'Description'     : 'Indicator for spam leads'
            , 'DisplayName'     : 'Spam Lead'
            , 'Category'        : 'Lead Information'
            , 'StatisticalType' : 'nominal'
            , 'Tags'            : 'Internal' }


def ds_company_isunusual(n):
    n = n.lower()

    if re.search('["#$%+:<=>?@\^_`{}~]', n):
        return 1

    if re.search('(^|\s+)[\[]*(none|no|not|delete|asd|sdf|unknown|undisclosed|null|dont|don\'t|n\/a|n\.a|abc|xyz|noname|nocompany)($|\s+)', n):
        return 1

    try:
        test_var = float(n)
        return 1
    except:
        pass

    return 0


def std_visidb_ds_spamindicator(firstName, lastName, title, phone, company):
    if firstName is None or lastName is None or title is None or phone is None or company is None:
        return None

    if std_visidb_ds_firstname_sameas_lastname(firstName, lastName) == 1:
        return 1

    score = 0
    score += ds_company_isunusual(company)

    if len(company) < 5:
        score += 1

    companyNameEntropy = std_visidb_ds_companyname_entropy(company)

    if companyNameEntropy is not None and companyNameEntropy <= 0.03:
        score += 1

    if len(title) <= 2:
        score += 1

    phoneEntropy = std_entropy(phone)

    if phoneEntropy is not None and phoneEntropy <= 0.03:
        score += 1

    if score >= 2:
        return 1

    return 0


def transform(args, record):
    column1 = args["column1"]
    column2 = args["column2"]
    column3 = args["column3"]
    column4 = args["column4"]
    column5 = args["column5"]
    firstName = record[column1]
    lastName = record[column2]
    title = record[column3]
    phone = record[column4]
    company = record[column5]
    return std_visidb_ds_spamindicator(firstName, lastName, title, phone, company)

File: le-transform/src/main/java/com/latticeengines/transform/v2_0_25/functions/StdVisidbDsSpamindicator.java
package com.latticeengines.transform.v2_0_25.functions;

import java.util.Map;
import java.util.regex.Pattern;

import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.latticeengines.transform.exposed.RealTimeTransform;
import com.latticeengines.transform.exposed.metadata.ApprovedUsage;
import com.latticeengines.transform.exposed.metadata.Category;
import com.latticeengines.transform.exposed.metadata.FundamentalType;
import com.latticeengines.transform.exposed.metadata.StatisticalType;
import com.latticeengines.transform.exposed.metadata.Tag;
import com.latticeengines.transform.exposed.metadata.TransformMetadata;

public class StdVisidbDsSpamindicator implements RealTimeTransform {

    private static final Logger log = LoggerFactory.getLogger(StdVisidbDsSpamindicator.class);

    private static final long serialVersionUID = 2222617772602907048L;

    public StdVisidbDsSpamindicator() {
    }

    private static Pattern pattern = Pattern.compile("(^|\\s+)[\\[]*(none|no|not|delete|"
            + "asd|sdf|unknown|undisclosed|" + "null|dont|don\'t|n/a|n.a" + "|abc|xyz|noname|nocompany)($|\\s+)");

    public StdVisidbDsSpamindicator(String modelPath) {
    }

    public Object transform(Map<String, Object> arguments, Map<String, Object> record) {
        String column1 = (String) arguments.get("column1");
        String column2 = (String) arguments.get("column2");
        String column3 = (String) arguments.get("column3");
        String column4 = (String) arguments.get("column4");
        String column5 = (String) arguments.get("column5");

        String firstName = column1 == null ? null : String.valueOf(record.get(column1));
        String lastName = column2 == null ? null : String.valueOf(record.get(column2));
        String title = column3 == null ? null : String.valueOf(record.get(column3));
        String phone = column4 == null ? null : String.valueOf(record.get(column4));
        String companyName = column5 == null ? null : String.valueOf(record.get(column5));

        if (firstName.equals("null"))
            firstName = "";
        if (lastName.equals("null"))
            lastName = "";
        if (title.equals("null"))
            title = "";
        if (phone.equals("null"))
            phone = "";
        if (companyName.equals("null"))
            companyName = "";

        return calculateStdVisidbDsSpamindicator(firstName, lastName, title, phone, companyName);
    }

    public static int calculateStdVisidbDsSpamindicator(String firstName, String lastName, String title, String phone,
            String companyName) {
        if (StdVisidbDsFirstnameSameasLastname.calcualteStdVisidbDsFirstnameSameasLastname(firstName, lastName))
            return 1;

        int score = 0;
        score += ds_company_isunusual(companyName);

        if (StdLength.calculateStdLength(companyName) < 5)
            score += 1;

        Double companyNameEntropy = StdVisidbDsCompanynameEntropy.calculateStdVisidbDsCompanynameEntropy(companyName);

        if (companyNameEntropy != null && companyNameEntropy <= 0.03)
            score += 1;

        if (StdLength.calculateStdLength(title) <= 2)
            score += 1;

        Double phoneEntropy = StdEntropy.calculateStdEntropy(phone);

        if (phoneEntropy != null && phoneEntropy <= 0.03)
            score += 1;

        if (score >= 2)
            return 1;

        return 0;
    }

    public static int ds_company_isunusual(String companyName) {
        if (companyName == null || StringUtils.isEmpty(companyName))
            return 0;

        companyName = companyName.toLowerCase();

        if (Pattern.matches(".*[\"#$%+:<=>?@\\^_`{}~].*", companyName))
            return 1;

        if (pattern.matcher(companyName).find())
            return 1;

        try {
            Float.valueOf(companyName);
            return 1;
        } catch (Exception ignore) {
            // companyName is not a float. good
        }

        return 0;
    }

    @Override
    public TransformMetadata getMetadata() {
        TransformMetadata metadata = new TransformMetadata();
        metadata.setApprovedUsage(ApprovedUsage.MODEL_MODELINSIGHTS);
        metadata.setCategory(Category.LEAD_INFORMATION);
        metadata.setDisplayName("Spam Lead");
        metadata.setFundamentalType(FundamentalType.BOOLEAN);
        metadata.setStatisticalType(StatisticalType.NOMINAL);
        metadata.setTags(Tag.INTERNAL_TRANSFORM);
        metadata.setDescription("Indicator for spam leads");
        return metadata;
    }
}
